File: backend/audit-service/app/models/audit.py
from sqlalchemy import String, DateTime, Integer, Text
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.sql import func
from sqlalchemy.orm import Mapped, mapped_column
import uuid
from datetime import datetime
from typing import Optional, Any
from app.database import Base

class AuditLog(Base):
    __tablename__ = "audit_logs"
    
    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    user_email: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    user_name: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    action: Mapped[str] = mapped_column(String(100), nullable=False)
    resource_type: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    resource_id: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    resource_name: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    details: Mapped[Optional[dict]] = mapped_column(JSONB, default=dict)
    ip_address: Mapped[Optional[str]] = mapped_column(String(45), nullable=True)
    user_agent: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(20), default="success")
    error_message: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    business_unit_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    organization_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    service_name: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    correlation_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    duration_ms: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": str(self.id),
            "user_id": str(self.user_id) if self.user_id else None,
            "user_email": self.user_email,
            "user_name": self.user_name,
            "action": self.action,
            "resource_type": self.resource_type,
            "resource_id": self.resource_id,
            "resource_name": self.resource_name,
            "details": self.details or {},
            "ip_address": self.ip_address,
            "user_agent": self.user_agent,
            "status": self.status,
            "error_message": self.error_message,
            "business_unit_id": str(self.business_unit_id) if self.business_unit_id else None,
            "organization_id": str(self.organization_id) if self.organization_id else None,
            "service_name": self.service_name,
            "correlation_id": str(self.correlation_id) if self.correlation_id else None,
            "duration_ms": self.duration_ms,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            # Computed user object for backward compatibility
            "user": {
                "id": str(self.user_id),
                "email": self.user_email,
                "username": self.user_email.split('@')[0] if self.user_email else None,
                "full_name": self.user_name
            } if self.user_id else None
        }

File: frontend/services/api/audit.ts
/**
 * Audit Logs API Service
 */
import { apiClient } from './client';

export interface AuditLog {
    id: string;
    user_id?: string;
    user_email?: string;
    user_name?: string;
    action: string;
    resource_type?: string;
    resource_id?: string;
    resource_name?: string;
    details?: Record<string, any>;
    ip_address?: string;
    user_agent?: string;
    status: string;
    error_message?: string;
    business_unit_id?: string;
    organization_id?: string;
    service_name?: string;
    correlation_id?: string;
    duration_ms?: number;
    created_at: string;
    user?: {
        id: string;
        email: string;
        username: string;
        full_name?: string;
        avatar_url?: string;
    };
}

export interface AuditLogListResponse {
    items: AuditLog[];
    total: number;
}

export interface AuditLogFilters {
    skip?: number;
    limit?: number;
    user_id?: string;
    action?: string;
    resource_type?: string;
    resource_id?: string;
    service_name?: string;
    business_unit_id?: string;
    start_date?: string;
    end_date?: string;
    search?: string;
    status?: 'success' | 'failure';
}

export interface AuditStats {
    period_days: number;
    total: number;
    success: number;
    failure: number;
    top_actions: Array<{ action: string; count: number }>;
}

/**
 * List audit logs with filtering and pagination
 */
export async function listAuditLogs(filters: AuditLogFilters = {}): Promise<AuditLogListResponse> {
    const params = new URLSearchParams();
    
    if (filters.skip !== undefined) params.append('skip', filters.skip.toString());
    if (filters.limit !== undefined) params.append('limit', filters.limit.toString());
    if (filters.user_id) params.append('user_id', filters.user_id);
    if (filters.action) params.append('action', filters.action);
    if (filters.resource_type) params.append('resource_type', filters.resource_type);
    if (filters.resource_id) params.append('resource_id', filters.resource_id);
    if (filters.service_name) params.append('service_name', filters.service_name);
    if (filters.business_unit_id) params.append('business_unit_id', filters.business_unit_id);
    if (filters.start_date) params.append('start_date', filters.start_date);
    if (filters.end_date) params.append('end_date', filters.end_date);
    if (filters.search) params.append('search', filters.search);
    if (filters.status) params.append('status', filters.status);
    
    const queryString = params.toString();
    const endpoint = `/api/v1/audit-logs${queryString ? `?${queryString}` : ''}`;
    
    return apiClient.request<AuditLogListResponse>(endpoint);
}

/**
 * Get a single audit log by ID
 */
export async function getAuditLog(logId: string): Promise<AuditLog> {
    return apiClient.request<AuditLog>(`/api/v1/audit-logs/${logId}`);
}

/**
 * Get audit log statistics
 */
export async function getAuditStats(days: number = 7, businessUnitId?: string): Promise<AuditStats> {
    const params = new URLSearchParams();
    params.append('days', days.toString());
    if (businessUnitId) params.append('business_unit_id', businessUnitId);
    
    return apiClient.request<AuditStats>(`/api/v1/audit-logs/stats/summary?${params.toString()}`);
}

export const auditApi = {
    listAuditLogs,
    getAuditLog,
    getAuditStats,
};
